File: data/convert_stbs_to_congress114.R
# Convert STBS 114th Congress data to congress109-style format.
# Requires: RcppCNPy
# Run from StatisticalModeling/data directory.

library(RcppCNPy)

temp_dir <- "temp_stbs"
out_dir <- "."

# Unzip npz to get npy files
if (!dir.exists(file.path(temp_dir, "npz_extracted"))) {
  dir.create(file.path(temp_dir, "npz_extracted"), recursive = TRUE)
  unzip(file.path(temp_dir, "counts114.npz"), exdir = file.path(temp_dir, "npz_extracted"))
}
npz_dir <- file.path(temp_dir, "npz_extracted")

# Load sparse matrix components (CSR format)
data_vec <- npyLoad(file.path(npz_dir, "data.npy"))
indices <- npyLoad(file.path(npz_dir, "indices.npy"))
indptr <- npyLoad(file.path(npz_dir, "indptr.npy"))
shape <- npyLoad(file.path(npz_dir, "shape.npy"))
n_rows <- as.integer(shape[1])
n_cols <- as.integer(shape[2])

# Build sparse matrix and convert to dense for aggregation
# CSR: row i has elements data[indptr[i]:(indptr[i+1]-1)] at columns indices[...]
counts <- matrix(0, n_rows, n_cols)
for (i in seq_len(n_rows)) {
  start <- indptr[i] + 1  # 0-based to 1-based
  end <- indptr[i + 1]
  if (end >= start) {
    idx <- start:end
    cols <- indices[idx] + 1  # 0-based to 1-based
    counts[i, cols] <- data_vec[idx]
  }
}

# Load author indices and author info
author_indices <- as.vector(npyLoad(file.path(temp_dir, "author_indices114.npy")))
author_info <- read.csv(file.path(temp_dir, "author_info114.csv"), row.names = 1)

# Ensure dimensions match
stopifnot(length(author_indices) == n_rows)

# Vocabulary - convert space to dot
vocab <- readLines(file.path(temp_dir, "vocabulary114.txt"))
vocab_dots <- gsub(" ", ".", vocab)

# Aggregate counts by author (author_indices is 0-based)
n_authors <- nrow(author_info)
author_counts <- matrix(0, n_authors, n_cols)
for (a in seq_len(n_authors)) {
  mask <- which(author_indices == (a - 1))
  if (length(mask) > 0) {
    author_counts[a, ] <- colSums(counts[mask, , drop = FALSE])
  }
}

# Member names from author_map (has "First Last (Party)" format)
author_map <- readLines(file.path(temp_dir, "author_map114.txt"))
names <- trimws(gsub(" \\([DIR]\\)$", "", author_map))

# congress114.csv - first column "name" for row.names when read by congress114.R
count_df <- data.frame(name = names, author_counts)
colnames(count_df)[-1] <- vocab_dots
write.csv(count_df, file.path(out_dir, "congress114.csv"), row.names = FALSE)

# congress114members.csv - row names must match congress114.csv
members <- data.frame(
  name = names,
  party = author_info$party,
  state = author_info$state,
  chamber = "S",
  repshare = NA_real_,
  cs1 = NA_real_,
  cs2 = NA_real_
)
rownames(members) <- names
write.csv(members, file.path(out_dir, "congress114members.csv"))

cat("Created congress114.csv:", nrow(count_df), "members x", ncol(count_df) - 1, "phrases\n")
cat("Created congress114members.csv:", nrow(members), "members\n")

File: data/convert_stbs_to_congress114.py
#!/usr/bin/env python3
"""Convert STBS 114th Congress data to congress109-style format.
Requires: numpy, scipy, pandas
Run from StatisticalModeling/data directory.
"""
import numpy as np
import pandas as pd
from scipy import sparse
from pathlib import Path

TEMP_DIR = Path("temp_stbs")
OUT_DIR = Path(".")

# Load data
counts = sparse.load_npz(TEMP_DIR / "counts114.npz")
author_indices = np.load(TEMP_DIR / "author_indices114.npy")
with open(TEMP_DIR / "vocabulary114.txt") as f:
    vocab = [line.strip() for line in f]
author_info = pd.read_csv(TEMP_DIR / "author_info114.csv", index_col=0)

# Convert vocabulary to dot format (congress109 uses "word.word" not "word word")
vocab_dots = [v.replace(" ", ".") for v in vocab]

# Aggregate counts by author: sum rows for each author
n_authors = author_info.shape[0]
n_terms = counts.shape[1]
author_counts = np.zeros((n_authors, n_terms))
for i in range(n_authors):
    mask = author_indices == i
    author_counts[i, :] = np.asarray(counts[mask, :].sum(axis=0)).flatten()

# Build member names (First Last format from author_info)
names = author_info["name"].str.title().values

# congress114.csv: rows=members, cols=phrases
count_df = pd.DataFrame(author_counts, index=names, columns=vocab_dots)
count_df.index.name = "name"
count_df.to_csv(OUT_DIR / "congress114.csv")

# congress114members.csv: name, party, state, chamber
# STBS is Senate only; format names to match congress109 (Title Case)
members = author_info.copy()
members["name"] = members["name"].str.title()
members["chamber"] = "S"  # Senate only
# congress109 has repshare, cs1, cs2 - we don't have these, use NA
members["repshare"] = np.nan
members["cs1"] = np.nan
members["cs2"] = np.nan
members = members[["name", "party", "state", "chamber", "repshare", "cs1", "cs2"]]
members.to_csv(OUT_DIR / "congress114members.csv", index=False)

print(f"Created congress114.csv: {count_df.shape[0]} members x {count_df.shape[1]} phrases")
print(f"Created congress114members.csv: {members.shape[0]} members")
